clippy_utils/src/source.rs

```diff
@@ -1,201 +1,602 @@
-        self
-    }
+
+/// A position in the `SourceMap` and the `SyntaxContext` it came from.
+#[derive(Clone, Copy)]
+pub struct PosWithCtxt {
+    pub pos: BytePos,
+    pub ctxt: SyntaxContext,
 }
 
-/// Conversion of a value into a `Span`
-pub trait IntoSpan: Sized {
-    fn into_span(self) -> Span;
-    fn with_ctxt(self, ctxt: SyntaxContext) -> Span;
+pub trait SpanLike: Copy {
+    /// Gets this value as an interned `Span`.
+    #[must_use]
+    fn span(self) -> Span;
+    /// Gets this `Span`'s data.
+    #[must_use]
+    fn data(self) -> SpanData;
 }
-impl IntoSpan for Span {
-    fn into_span(self) -> Span {
+impl SpanLike for Span {
+    #[inline]
+    fn span(self) -> Span {
         self
     }
-    fn with_ctxt(self, ctxt: SyntaxContext) -> Span {
-        self.with_ctxt(ctxt)
+    #[inline]
+    fn data(self) -> SpanData {
+        self.data()
     }
 }
-impl IntoSpan for SpanData {
-    fn into_span(self) -> Span {
-        self.span()
+impl SpanLike for SpanData {
+    #[inline]
+    fn span(self) -> Span {
+        Span::new(self.lo, self.hi, self.ctxt, self.parent)
     }
-    fn with_ctxt(self, ctxt: SyntaxContext) -> Span {
-        Span::new(self.lo, self.hi, ctxt, self.parent)
+    #[inline]
+    fn data(self) -> SpanData {
+        self
     }
 }
-impl IntoSpan for Range<BytePos> {
-    fn into_span(self) -> Span {
-        Span::with_root_ctxt(self.start, self.end)
+impl SpanLike for &SpanData {
+    #[inline]
+    fn span(self) -> Span {
+        Span::new(self.lo, self.hi, self.ctxt, self.parent)
     }
-    fn with_ctxt(self, ctxt: SyntaxContext) -> Span {
-        Span::new(self.start, self.end, ctxt, None)
+    #[inline]
+    fn data(self) -> SpanData {
+        *self
     }
 }
 
-pub trait SpanRangeExt: SpanRange {
-    /// Attempts to get a handle to the source text. Returns `None` if either the span is malformed,
-    /// or the source text is not accessible.
-    fn get_source_text(self, cx: &impl HasSession) -> Option<SourceText> {
-        get_source_range(cx.sess().source_map(), self.into_range()).and_then(SourceText::new)
+// Workaround for `array::map` not working with `track_caller`. This will fail to drop values if a
+// panic occurs, but that doesn't matter for our use.
+const fn mk_uninit<T, U, const N: usize>(_: &[T; N]) -> [MaybeUninit<U>; N] {
+    [const { MaybeUninit::uninit() }; N]
+}
+macro_rules! map_array_inline {
+    ($a:expr, $p:pat => $e:expr $(,)?) => {{
+        let src = ManuallyDrop::new($a);
+        let mut dst = mk_uninit(&src);
+        for i in 0..dst.len() {
+            // SAFETY: `src` was fully initialized at the start of the loop.
+            let $p = unsafe { (&raw const src[i]).read() };
+            dst[i].write($e);
+        }
+        // SAFETY: The loop has written every element in the array.
+        unsafe { MaybeUninit::array_assume_init(dst) }
+    }};
+}
+
+/// Helper functions to interact with the source text of a span/range.
+pub trait SpanExt: Sized {
```

Comment:
Could do with a `Result`-style `See the module documentation for details.` link here